esquery.js

```diff
@@ -47,196 +46,267 @@ function getPath(obj, key) {
+            const nth = selector.index.value;
+            return (node, ancestry, options) => {
+                return matches(node, selector.right, ancestry, options) &&
+                    nthChild(node, ancestry, nth, true, options);
+            };
+        }
+
+        case 'class': {
+            const name = selector.name.toLowerCase();
+            return (node, ancestry) => {
+                switch(name){
+                    case 'statement':
+                        if(node.type.slice(-9) === 'Statement') return true;
+                        // fallthrough: interface Declaration <: Statement { }
+                    case 'declaration':
+                        return node.type.slice(-11) === 'Declaration';
+                    case 'pattern':
+                        if(node.type.slice(-7) === 'Pattern') return true;
+                        // fallthrough: interface Expression <: Node, Pattern { }
+                    case 'expression':
+                        return node.type.slice(-10) === 'Expression' ||
+                            node.type.slice(-7) === 'Literal' ||
+                            (
+                                node.type === 'Identifier' &&
+                                (ancestry.length === 0 || ancestry[0].type !== 'MetaProperty')
+                            ) ||
+                            node.type === 'MetaProperty';
+                    case 'function':
+                        return node.type === 'FunctionDeclaration' ||
+                            node.type === 'FunctionExpression' ||
+                            node.type === 'ArrowFunctionExpression';
+                }
+                throw new Error(`Unknown class name: ${selector.name}`);
+            };
+        }
     }
 
     throw new Error(`Unknown selector type: ${selector.type}`);
 }
 
+/**
+ * @callback TraverseOptionFallback
+ * @param {external:AST} node The given node.
+ * @returns {string[]} An array of visitor keys for the given node.
+ */
+/**
+ * @typedef {object} ESQueryOptions
+ * @property { { [nodeType: string]: string[] } } [visitorKeys] By passing `visitorKeys` mapping, we can extend the properties of the nodes that traverse the node.
+ * @property {TraverseOptionFallback} [fallback] By passing `fallback` option, we can control the properties of traversing nodes when encountering unknown nodes.
+ */
+
+/**
+ * Given a `node` and its ancestors, determine if `node` is matched
+ * by `selector`.
+ * @param {?external:AST} node
+ * @param {?SelectorAST} selector
+ * @param {external:AST[]} [ancestry=[]]
+ * @param {ESQueryOptions} [options]
+ * @throws {Error} Unknowns (operator, class name, selector type, or
+ * selector value type)
+ * @returns {boolean}
+ */
+function matches(node, selector, ancestry, options) {
+    if (!selector) { return true; }
+    if (!node) { return false; }
+    if (!ancestry) { ancestry = []; }
+
+    if (!selector._match) {
+        selector._match = createMatcher(selector);
```

Comment:
I don't love this change. `selector` comes in from outside and I don't really feel that it's appropriate for us to modify it. If I had done this myself, I probably would've used a `WeakMap`. How much do we care about this particular performance improvement being available for older engines that don't have `WeakMap`? I'm generally okay with asking users to upgrade if they are looking for better performance.

Reply:
Also I would probably move the memoisation into `createMatcher`.

Reply:
Just to avoid back and forth on the PR review, is this comment asking to make a change or just sharing thoughts out loud? As is, these changes already significantly improve performance of esquery. It would be nice to split preference/and engine support conversations to a separate PR, after this one.

Reply:
This comment was asking for any arguments for why we might *not* want to take the suggestion of switching it to a `WeakMap`. If we don't have a good reason, I would like to make that change. Don't worry, it's just a few line diff and shouldn't take more than a minute.

Reply:
Thank you